query/src/main/java/org/infinispan/query/QueryDefinition.java

```diff
@@ -0,0 +1,104 @@
+package org.infinispan.query;
+
+import java.util.Optional;
+
+import org.apache.lucene.search.Filter;
+import org.apache.lucene.search.Sort;
+import org.hibernate.search.filter.FullTextFilter;
+import org.hibernate.search.query.engine.spi.HSQuery;
+import org.infinispan.AdvancedCache;
+import org.infinispan.query.dsl.embedded.impl.EmbeddedQueryEngine;
+import org.infinispan.query.dsl.embedded.impl.HsQueryRequest;
+import org.infinispan.query.dsl.embedded.impl.QueryEngine;
+import org.infinispan.query.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+/**
+ * Stores the query to be executed in a cache in either a String or {@link HSQuery} form together with pagination
+ * and sort information.
+ *
+ * @since 9.2
+ */
+public class QueryDefinition {
+
+   private static final Log log = LogFactory.getLog(QueryDefinition.class, Log.class);
+
+   private String queryString;
+   private HSQuery hsQuery;
+   private int maxResults = 100;
+   private int firstResult;
+
+   private transient Sort sort;
+
+   public QueryDefinition(String queryString) {
+      this.queryString = queryString;
+   }
+
+   public QueryDefinition(HSQuery hsQuery) {
+      this.hsQuery = hsQuery;
+   }
+
+   public Optional<String> getQueryString() {
+      return Optional.ofNullable(queryString);
+   }
+
+   public HSQuery getHsQuery(AdvancedCache<?, ?> cache) {
+      if (hsQuery == null) {
+         QueryEngine queryEngine = cache.getComponentRegistry().getComponent(EmbeddedQueryEngine.class);
```

Comment:
Would it be possible to pass the EmbeddedQueryEngine somehow to QueryDefinition instead of grabbing it from the component registry? Maybe in the ``` QueryDefinition(String queryString)``` constructor.

Reply:
Unfortunately no, the ```QueryDefinition``` is what is broadcast, so the serialization layers is responsible to construct it.